main/src/hadd.cxx

```diff
@@ -119,252 +134,335 @@
          } else {
-            Bool_t hasFollowupNumber = kTRUE;
-            for (char *c = argv[a+1]; *c != '\0'; ++c) {
-               if (!isdigit(*c)) {
-                  // Verbosity level was not specified use the default:
-                  hasFollowupNumber = kFALSE;
-                  break;
-               }
-            }
-            if (hasFollowupNumber) {
-               Long_t request = strtol(argv[a+1], 0, 10);
-               if (request < kMaxLong && request >= 0) {
-                  verbosity = (Int_t)request;
-                  ++a;
-                  ++ffirst;
-               } else {
-                  verbosity = 99;
-                  std::cerr << "Error: could not parse the verbosity level passed after -v: " << argv[a+1] << ". We will use the default value (99).\n";
-               }
-            }
-         }
-         ++ffirst;
-      } else if ( argv[a][0] == '-' ) {
-         bool farg = false;
-         if (force && argv[a][1] == 'f') {
-            // Bad argument
-            std::cerr << "Error: Using option " << argv[a] << " more than once is not supported.\n";
-            ++ffirst;
-            farg = true;
-         }
-         const char *prefix = "";
-         if (argv[a][1] == 'f' && argv[a][2] == 'k') {
-            farg = true;
-            force = kTRUE;
-            keepCompressionAsIs = kTRUE;
-            prefix = "k";
-         }
-         if (argv[a][1] == 'f' && argv[a][2] == 'f') {
-            farg = true;
-            force = kTRUE;
-            useFirstInputCompression = kTRUE;
-            if (argv[a][3] != '\0') {
-               std::cerr << "Error: option -ff should not have any suffix: " << argv[a] << " (suffix has been ignored)\n";
-            }
-         }
-         char ft[7];
-         for (int alg = 0; !useFirstInputCompression && alg <= 5; ++alg) {
-            for( int j=0; j<=9; ++j ) {
-               const int comp = (alg*100)+j;
-               snprintf(ft,7,"-f%s%d",prefix,comp);
-               if (!strcmp(argv[a],ft)) {
-                  farg = true;
-                  force = kTRUE;
-                  newcomp = comp;
-                  break;
-               }
-            }
+            std::cerr << "[err] Invalid flag: " << arg << "\n";
+            return EFlagResult::Error;
          }
-         if (!farg) {
-            // Bad argument
-            std::cerr << "Error: option " << argv[a] << " is not a supported option.\n";
-         }
-         ++ffirst;
-      } else if (!outputPlace) {
-         outputPlace = a;
       }
+      ++cur;
+   }
+   
+   return EFlagResult::Parsed;
+}
+
+// Returns nullopt if any of the flags failed to parse.
+// If an unknown flag is encountered, it will print a warning and go on.
+static std::optional<HaddArgs> ParseArgs(int argc, char **argv)
+{
+   HaddArgs args {};
+
+   for (int argIdx = 1; argIdx < argc; ++argIdx) {
+      const char *argRaw = argv[argIdx];
+      if (argRaw[0] == '-' && argRaw[1] != '\0') {
```

Comment:
Can we be sure that `argRaw[1]` exists, i.e. that `argRaw[0]` is not `\0`?

Reply:
I'm pretty sure that cannot happen under "normal" conditions, but I'll add the check for good measure